Mill 3. gen add support for PRECISION_HALVES

LaStrada opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add support for halves precision (`PRECISION_HALVES`) for 3. generation heaters from Mill when using local API. I sent them an email and they confirmed that only 3. gen supports this. First and second only support `PRECISION_WHOLE`.

Email from Mill (Norwegian):
> For å svare på det første spørsmålet, så er det kun gen 3 som støtter halve grader. Gen 2-ovner vil runde ned til nærmeste hele grad, så 22.5 blir f.eks 22 på gen 2.

Google translate:
> To answer the first question, only gene 3 supports half degrees. Gen 2 ovens will round down to the nearest whole degree, so 22.5 will be 22 on gen 2, for example.

To be able to do the same thing for cloud, the `pymill` library needs to be updated to support `float` values, but that can be done later if this is needed.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: N/A
- This PR is related to issue: N/A
- Link to documentation pull request: N/A

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

LaStrada committed:
Add support for PRECISION_HALVES for Mill 3. generation heaters

LaStrada committed:
Add support for PRECISION_HALVES for Mill 3. generation heaters

LaStrada committed:
Add support for precision halves for Mill local API

LaStrada committed:
Add support for precision halves for Mill local API

LaStrada committed:
Make sure to cast to float for local api

LaStrada committed:
Make sure to cast to float for local api

LaStrada committed:
Support both float for gen 3 heaters and int for gen < 3 heaters when using cloud api

LaStrada committed:
Support both float for gen 3 heaters and int for gen < 3 heaters when using cloud api

LaStrada committed:
Mill attribute bugfix

LaStrada committed:
Mill attribute bugfix

LaStrada commented on homeassistant/components/mill/climate.py:
```diff
@@ -119,15 +121,20 @@ def __init__(self, coordinator, heater):
         if (temperature := kwargs.get(ATTR_TEMPERATURE)) is None:
             return
+        if self._generation == 3:
+            temperature = float(temperature)
```
I'm not sure if this is needed. Can I use float for all of the generations? As I said in the description, I don't have a first or second gen heater, so I can't verify. But I belive the API supports `float` for all generations, but is rouded down for < 3. gen.

gjohansson-ST replied:
If you set the precision correctly I think you can use the temperature helper https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/temperature.py#L11

LaStrada committed:
Merge branch 'dev' into mill-3-gen-add-precision-halves

gjohansson-ST commented on homeassistant/components/mill/climate.py:
```diff
@@ -200,7 +207,7 @@ class LocalMillHeater(CoordinatorEntity, ClimateEntity):
     _attr_min_temp = MIN_TEMP
     _attr_supported_features = ClimateEntityFeature.TARGET_TEMPERATURE
-    _attr_target_temperature_step = PRECISION_WHOLE
+    _attr_target_temperature_step = PRECISION_HALVES
```
```suggestion
    _attr_target_temperature_step = PRECISION_HALVES if self._generation == 3 else PRECISION_WHOLE
```
If you do like this then I think it's enough to use the old code in set_temp but with float instead of int (as float without decimal will be returned as int)

LaStrada replied:
Looks good @gjohansson-ST ! Will test this out later (been on vacation for a while).

If I understands you correctly I can revert all the changes I did in `async_set_temperature`?

LaStrada replied:
Local API for Mill requires gen 3 heaters, and `PRECISION_HALVES` is always supported. So there is no need for `PRECISION_WHOLE` here.

LaStrada replied:
But tested this for the cloud integration (line 93), but I'm getting this error:
```NameError: name 'self' is not defined```

LaStrada committed:
Merge branch 'dev' into mill-3-gen-add-precision-halves

LaStrada committed:
Merge branch 'home-assistant:dev' into mill-3-gen-add-precision-halves

LaStrada committed:
Revert PRECISION_HALVES for Mill cloud integration

LaStrada committed:
Merge branch 'mill-3-gen-add-precision-halves' of github.com:LaStrada/core into mill-3-gen-add-precision-halves

homeassistant commented:
Hi @GitHub-Action, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#73592).

Once you do that we will be able to review and accept this pull request.

Thanks!

LaStrada committed:
Remove unused code

homeassistant commented:
Hi @GitHub-Action, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#73592).

Once you do that we will be able to review and accept this pull request.

Thanks!

LaStrada committed:
Merge branch 'dev' into mill-3-gen-add-precision-halves

LaStrada commented:
I updated the PR and the description to reflect the latest changes I made.

TLDR: Reverted to only use `PRECISION_WHOLE` for cloud since there are some issues there. Need to update `pymill` to use `float` instead of `int` for generation 3 heaters.

LaStrada committed:
Revert to casting to int for mill cloud integration

LaStrada committed:
Merge branch 'mill-3-gen-add-precision-halves' of github.com:LaStrada/core into mill-3-gen-add-precision-halves

LaStrada committed:
Remove unused code

allenporter reviewed (approved):
Looks good to me. (I reviewed the dev docs, and couple other integrations using this and it seems consistent/fine).
